docs(admin): revise the brute force protection configuration chapter (#11541)
nextcloud/documentation: admin_manual/configuration_server/bruteforce_configuration.rst

The primary filter is IP address-based. This means that any account - even one associated with
a given brute force attempt - is not impacted when it is connecting from a different IP address
than any brute force attempts. This helps minimize inadvertent denial of service attacks against 
legitimate connections, while maximizing attack resistance from problematic IP sources.

Nuisance triggers are minimized through reasonable built-in defaults appropriate to each type of action.

The attempts history is automatically managed by a daily cronjob. Individual entries
expire after 48 hours (attempts, however, may be still *logged* indefinitely elsewhere through the usual 
mechanisms within Nextcloud Server and at the discretion of the admin).

Excluding (whitelisting) select IP addresses from brute force protection to prevent false 
positives is supported, but usually false positives are best handled by fixing the underlying causes 
(e.g. a misconfigured reverse proxy or misbehaving client). 

.. tip:: If you do notice a problem with the authentication behavior of any the official Nextcloud clients, 

>    consider report it to the appropriate repository so that it can be looked into.

Author: ```suggestion
  please report it to the appropriate repository so that it can be looked into.
```

Keeping brute force protection active and operating properly helps protects your Nextcloud Server from 
malicious actors while minimizing potential impact on legitimate usage.

Example: The login page
~~~~~~~~~~~~~~~~~~~~~~~

The brute force protection is easiest to see in action on the login page.
If you try to log in the first time with an invalid username and/or password you
will not notice anything. But if you do this a few times you start to notice
that the verification of the login is taking longer each time. This is the
brute force protection kicking in.

The maximum delay is 25 seconds, unless maximum number of attempts (currently 10) was reached within 
the last 30 minutes (in which case a ``429 Too Many Requests`` will be returned until the maximum attempts 
within the most recent 30 minutes has dropped below the threshold).

After a successful login (from the same source IP address), any prior invalid login attempts will be cleared 
and you will no longer be hit by the delay.

> .. note:: Not all actions are necessarily viewed the same. It is possible for some activities to be more strict

Author: ```suggestion
.. note:: Not all actions are necessarily viewed the same. It is possible for some activities to be more (or less) strict
```

you can see which IP addresses are throttled and any other metadata stored about their attempts to 
connect.

Excluding IP addresses from brute force protection
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note:: Most nuisance triggering of brute force protection can be resolved through proper configuration of reverse 
   proxies. In other cases, select IP addresses that need to be whitelisted can be configured within this app (while 
   leaving brute force protection enabled). This can be useful for testing purposes or when there are a lot of people 
   (or devices) connecting from a known, single IP address.

It's possible to exclude IP addresses from the brute force protection.

- Make sure the ``bruteforcesettings`` app is enabled (it is by default)
- Login as admin and go to **Administration settings -> Security**

.. danger::

   Any excluded IP address can perform authentication attempts without any throttling.
   It's best to exclude as few IP addresses as you can, or even none at all. 

Additional Details
------------------

Feature History
~~~~~~~~~~~~~~~

(Notable)

> * Server 10.0:
>    - Added: Initial implementation
>    - Changed: Protection of major authentication entrypoints
> * Server 12.0:
>    - Added: Annotation support for protecting API calls
>    - Added: Distinct action types for more complex handling combinations
>    - Changed: Protection of some public APIs
>    - Changed: Protection of public links
>    - Fixed: Disregard existing recent attempts immediately if BPF is disabled manually
>    - Added: Admin security settings section for managing some parameters (via the ``bruteforcesettings`` shipped app)
>    - Added: Enable the admin to exclude IP addresses from throttling (via the ``bruteforcesettings`` shipped app)
>    - Changed: Controllers now must explicly call throttle() to increase throttle counter
> * Server 13.0
>    - Changed: Reset bruteforce attempts upon successful login (only the entries associated with that user from that IP)
>    - Added: Make bruteforce status available via Capabilities
> * Server 14.0
>    - Fixed: Truncate long usernames (>64 characters) during failed login attempts before storing the attempt entry
>    - Fixed: Decrease maxDelay from 30s to 25s to avoid unintentional client authentication timeouts
> * Server 15.0
>    - Changed: Reset bruteforce attempts upon successful token refresh when using OAuth
> * Server 18.0
>    - Fixed: Handle scoped IPv6 addresses
> * Server 20.0
>    - Changed: When maximum delay is reached and maximum attempts permitted have been exceeded within the past 30 minutes, return "429 Too Many Requests" until no longer the true
>    - Added: ``occ`` command to reset bruteforce attempts for an IP (``occ security:bruteforce:reset``)
> * Server 21.0
>    - Added: Once a day cronjob added to cleanup stale attempt entries
>    - Changed: Delete all entries >48 hours old
> * Server 24.0
>    - Added: Protection of password reset page
>    - Added: Logging of throttling or blocking (info level)
> * Server 25.0
>    - Added: Enable apps to reset the delay for an IP
>    - Added: Enable apps to get the current delay for an IP
> * Server 26.0
>    - Changed: Protection for public shares
> * Server 28.0
>    - Fixed: Reset bruteforce attempts upon successful sudo attempt
>    - Fixed: CSP errors when injecting ``X-Nextcloud-Bruteforce-Throttled`` (the throttle delay header)
>    - Added: Memcache based backend
>    - Added: Show admins when they are throttled
>    - Added: Show current user throttling status (via the ``bruteforcesettings`` app)
> * Server 29.0
>    - Fixed: Don't throw a "500 Internal Server Error" when MaxDelayReached; instead return a "429 Too Many Requests"
>    - Fixed: Prevent setting empty IP masks when adding an exclusion (in the ``bruteforcesetting`` app) 

Reviewer: Not a fan of this block. It gives way too many details for attack vectors on old instances:

```suggestion
* Server 10:
   - Added: Initial implementation - Protecting major authentication entrypoints
* Server 12:
   - Added: Annotation support for protecting API calls
   - Added: Distinct action types for more complex handling combinations
   - Added: Enable the admin to exclude IP addresses from throttling (via the ``bruteforcesettings`` shipped app)
   - Changed: Controllers now must explicly call throttle() to increase throttle counter
* Server 13
   - Changed: Reset bruteforce attempts upon successful login (only the entries associated with that user from that IP)
   - Added: Make bruteforce status available via Capabilities
* Server 14
   - Fixed: Decrease maxDelay from 30s to 25s to avoid unintentional client authentication timeouts
* Server 18
   - Fixed: Handle scoped IPv6 addresses
* Server 20
   - Changed: When maximum delay is reached and maximum attempts permitted have been exceeded within the time frame, return "429 Too Many Requests"
   - Added: ``occ`` command to reset bruteforce attempts for an IP (``occ security:bruteforce:reset``)
* Server 24
   - Added: Logging of throttling or blocking (info level)
* Server 25
   - Added: Enable apps to reset the delay for an IP
   - Added: Enable apps to get the current delay for an IP
* Server 28
   - Added: Memcache based backend
   - Added: Show admins when they are throttled
   - Added: Show current IP throttling status (via the ``bruteforcesettings`` app)
```

but even then it is a mix of addressing admins and developers.

Author: Streamlined the history block: admin-only stuff now (removed as much as possible of the other + dev-only bits)
